### PR TITLE
Refactor Mobile Auth Strategy

### PR DESCRIPTION
**Pull request by DanielToby**
This PR should be in sync with the following PRs in imodel02 and mobile-sdk-ios:
[Narrow the iModelJsMobile auth API](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodel02/pullrequest/244115)
[Update ITMAuthorizationClient to conform to the new protocol](https://github.com/iTwin/mobile-sdk-ios/pull/55)

Mobile auth has been refactored such that logic like sign-in, sign-out and token fetching now take place natively. Developers on iOS are expected to consume the itwin-mobile-sdk Swift package and use `ITMAuthorizationClient` if they need auth. 
The authorization interface in iModelJsMobile has been reduced to a single function: `authGetAccessToken`. The same change has been made to the AuthorizationClient protocol. From TypeScript, the mobile auth strategy becomes simpler; if auth is configured / initialized, getAccessToken will return an access token. If for any reason it is not, an empty string is returned. In TypeScript, the mobile backend will request and cache the token provided by `authGetAccessToken`, and the mobile frontend will request and cache tokens by communicating with said backend via IPC.

`authAccessTokenChanged` is used to invalidate any cached access tokens when a change in state occurs.

Developers can no longer pass `authorizationClient` in their iModelApp config on iOS. If third party auth is required, they should define their auth client in Swift / Objective-C and conform to the AuthorizationClient protocol defined in iModelJsMobile. 

Similar changes will need to take place on Android for compatibility with these updates.

[Corresponding native PR](https://dev.azure.com/bentleycs/iModelTechnologies/_git/imodel02/pullrequest/244115).

### CODE REVIEW

File: core/mobile/src/backend/MobileAuthorizationBackend.ts [hunk 1]
```diff
@@ -6,159 +6,30 @@
...
     return new Promise<AccessToken>((resolve, reject) => {
-      MobileHost.device.authGetAccessToken((tokenStringJson?: AccessToken, err?: string) => {
-        if (!err && tokenStringJson) {
-          resolve(tokenStringJson);
-        } else {
-          reject(new Error(err));
-        }
-      });
+      if (this._accessToken) {
+        resolve(this._accessToken);
```
Comment (tcobbs-bentley) [inline core/mobile/src/backend/MobileAuthorizationBackend.ts:22]: This will never fetch a new token once the old one has expired.

My advice in or call from weeks ago was to update MobileHost.device.authGetAccessToken (called below) to include the expiration date as a new parameter in the result. Technically, the code can parse the token to extract the expiration date, but I feel that that is a bad idea.

Once you know the expiration date, always fetch a new token if the current time is within 1 minute of the expiration date.

File: core/mobile/src/frontend/MobileAuthorizationFrontend.ts [hunk 2]
```diff
@@ -0,0 +1,25 @@
...
+import { MobileApp } from "./MobileApp";
+
+/** Utility to request OIDC/OAuth tokens from mobile frontend to mobile backend.
+ * @beta
+ */
+export class MobileAuthorizationFrontend implements AuthorizationClient {
+  protected _accessToken: AccessToken = "";
+
+  public async getAccessToken(): Promise<AccessToken> {
+    if (!this._accessToken) {
```
Comment (tcobbs-bentley) [inline core/mobile/src/frontend/MobileAuthorizationFrontend.ts:20]: This will also continue to return the token after it has expired. You can either not cache it at all (which would cause it to always call the backend), or you need to update the getAccessToken call to the backend to return both the token and the expiration date. If you cache (which I think you should), then as with the backend, if the current time is within 1 minute of the expiration date, fetch a new one.

File: core/mobile/src/backend/MobileAuthorizationBackend.ts (hunk 1 above)
Comment (DanielToby) [reply core/mobile/src/backend/MobileAuthorizationBackend.ts:33]: Fixed, though I used ten minutes as the expiration buffer. This matches the electron auth behavior.

File: core/mobile/src/frontend/MobileAuthorizationFrontend.ts (hunk 2 above)
Comment (DanielToby) [reply core/mobile/src/frontend/MobileAuthorizationFrontend.ts]: Fixed.

File: core/mobile/src/backend/MobileAuthorizationBackend.ts [hunk 3]
```diff
@@ -14,21 +14,37 @@ import { MobileHost } from "./MobileHost";
...
+        this._fetchingToken = true;
+        MobileHost.device.authGetAccessToken((tokenString?: AccessToken, expirationDate?: String, error?: String) => {
           if (error) {
             this._accessToken = "";
             reject(error);
           }
+
           this._accessToken = tokenString ?? "";
+          if (expirationDate !== undefined)
+            this._expirationDate = new Date(expirationDate.toString())
```
Comment (tcobbs-bentley) [inline core/mobile/src/backend/MobileAuthorizationBackend.ts:44]: expirationDate is already a string, so .toString() is redundant.

Comment (DanielToby) [reply core/mobile/src/backend/MobileAuthorizationBackend.ts]: Fixed.

**Review by tcobbs-bentley — APPROVED** [review]

File: core/mobile/src/backend/MobileHost.ts [hunk 4]
```diff
@@ -153,10 +165,33 @@ export class MobileHost {
 
   /** Start the backend of a mobile app. */
   public static async startup(opt?: MobileHostOpts): Promise<void> {
+    const authorizationClient = new MobileAuthorizationBackend();
```
Comment (calebmshafer) [inline core/mobile/src/backend/MobileHost.ts:168]: I don't think you'll want to create and set this prior to the device being setup. I don't think you immediately try to use it but could lead to weird issues.

File: core/mobile/src/frontend/MobileAuthorizationFrontend.ts [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+/** @packageDocumentation
+ * @module OIDC
+ */
+
+import { AccessToken } from "@itwin/core-bentley";
+import { AuthorizationClient } from "@itwin/core-common";
+import { MobileApp } from "./MobileApp";
+
+/** Utility to request OIDC/OAuth tokens from mobile frontend to mobile backend.
+ * @beta
```
Comment (calebmshafer) [inline core/mobile/src/frontend/MobileAuthorizationFrontend.ts:14]: This should be `@internal` similar to the backend one.

File: test-apps/display-test-app/src/frontend/App.ts [hunk 6]
```diff
@@ -251,7 +251,7 @@ export class DisplayTestApp {
     if (ProcessDetector.isElectronAppFrontend) {
       await ElectronApp.startup(opts);
     } else if (ProcessDetector.isIOSAppFrontend) {
-      await IOSApp.startup(opts);
+      await IOSApp.startup(opts as IOSAppOpts);
```
Comment (calebmshafer) [inline test-apps/display-test-app/src/frontend/App.ts:254]: Why did you need to cast this?

**Comment by calebmshafer** [comment]
Does the display test app sign-in currently work or is that still broken? I didn't see the update to use the Swift side auth sign-in.

File: core/mobile/src/backend/MobileAuthorizationBackend.ts [hunk 7]
```diff
@@ -6,159 +6,51 @@
...
-   * Time in seconds that's used as a buffer to check the token for validity/expiry.
-   * The checks for authorization, and refreshing access tokens all use this buffer - i.e., the token is considered expired if the current time is within the specified
-   * time of the actual expiry.
-   * @note If unspecified this defaults to 10 minutes.
-   */
-  readonly expiryBuffer?: number;
-}
-
 /** Utility to provide OIDC/OAuth tokens from native ios app to frontend
  * @beta
```
[the remark this answers is not in the document]
Comment (DanielToby) [reply core/mobile/src/backend/MobileAuthorizationBackend.ts]: Fixed.

File: core/mobile/src/frontend/MobileAuthorizationFrontend.ts (hunk 5 above)
Comment (DanielToby) [reply core/mobile/src/frontend/MobileAuthorizationFrontend.ts]: Fixed.

File: test-apps/display-test-app/src/frontend/App.ts (hunk 6 above)
Comment (DanielToby) [reply test-apps/display-test-app/src/frontend/App.ts:254]: This is because I updated IOSAppOpts to prohibit the passing of an authorizationClient. Even though one isn't provided here, the cast can't be done implicitly.

Comment (tcobbs-bentley) [reply test-apps/display-test-app/src/frontend/App.ts:254]: Wasn't that change supposed to get moved to MobileAppOpts, with IOSAppOpts deleted?

Comment (DanielToby) [reply test-apps/display-test-app/src/frontend/App.ts:254]: Yes, though I was planning on doing that in a separate PR. @calebmshafer, should I instead include those changes here?

Comment (pmconne) [reply test-apps/display-test-app/src/frontend/App.ts:254]: Is it needed for 3.2 release?

**Review by calebmshafer — APPROVED** [review]

File: core/mobile/src/backend/MobileHost.ts [hunk 8]
```diff
@@ -195,8 +193,10 @@ export class MobileHost {
       // following will provide impl for device specific api.
       setupMobileRpc();
     }
+    const iModelHostConfiguration = opt?.iModelHost ?? new IModelHostConfiguration();
+    iModelHostConfiguration.authorizationClient = authorizationClient;
```
Comment (tcobbs-bentley) [inline core/mobile/src/backend/MobileHost.ts:197]: Is this a fix for the backend auth failure Todd was seeing?

Comment (DanielToby) [reply core/mobile/src/backend/MobileHost.ts:197]: Yes.
